monkey/agent_plugins/exploiters/smb/src/smb_exploit_client.py

```diff
@@ -1,264 +1,177 @@
+
+
+class SMBExploitClient(IRemoteAccessClient):
     """Manages the SMB connection, Exploitation events"""
 
     def __init__(
-        self, agent_event_publisher: IAgentEventPublisher, host: TargetHost, options: SMBOptions
+        self,
+        host: TargetHost,
+        options: SMBOptions,
+        smb_client: SMBClient,
     ):
-        self._agent_event_publisher = agent_event_publisher
         self._host = host
         self._options = options
-        self._smb: Optional[SMBConnection] = None
-        self._copied_file_details: Optional[CopiedFileDetails] = None
-        self._authenticated_credentials: Optional[Credentials] = None
+        self._smb_client = smb_client
 
-    def authenticate(self, credentials: Credentials) -> bool:
-        """Returns True if authentication succeeded, False otherwise
-        Side effect: The SMB connection is established on success"""
+    def login(self, credentials: Credentials, tags: Set[str]):
+        tags.update(LOGIN_TAGS)
 
-        self._smb = create_smb_connection(self._host)
-        if not self._smb:
-            return False
-
-        if not smb_login(self._smb, credentials):
-            return False
-
-        self._smb.setTimeout(self._options.smb_connect_timeout)
-
-        if logout_guest(self._smb):
-            return False
+        try:
+            self._smb_client.connect_with_user(
+                self._host, credentials, timeout=self._options.smb_connect_timeout
+            )
+        except Exception as err:
+            error_message = f"Failed to authenticate over SMB with {credentials}: {err}"
+            raise RemoteAuthenticationError(error_message)
 
-        self._authenticated_credentials = credentials
-        return True
+    def _raise_if_not_authenticated(self, error_type: Type[Exception]):
+        if not self._smb_client.connected():
+            raise error_type(
+                "This operation cannot be performed until authentication is successful"
+            )
 
     def get_os(self) -> OperatingSystem:
         return OperatingSystem.WINDOWS
 
-    def execute(
-        self,
-        servers: Sequence[str],
-        current_depth: int,
-        interrupt: Event,
-    ):
-        """Raises an exception if the execution failed"""
-        if not self._authenticated_credentials:
-            raise Exception("Not authenticated")
-        rpc = self._connect_rpc(self._authenticated_credentials, self._options.smb_connect_timeout)
-
-        if not self._copied_file_details:
-            raise Exception("File was not copied before executing it")
-        command = build_smb_command(
-            servers,
-            current_depth,
-            self._copied_file_details.remote_path,
-            self._copied_file_details.destination_path,
-        )
-
-        rpc.bind(scmr.MSRPC_UUID_SCMR)
-        resp = scmr.hROpenSCManagerW(rpc)
-        sc_handle = resp["lpScHandle"]
+    def execute_detached(self, command: str, tags: Set[str]):
```

Comment:
Will it always be possible to create a detached process (for brute force exploiters in general)?

Reply:
From the `BruteForceExploiter's` perspective, it's a necessity that the command be detached. How that gets achieved is up to the client.

@cakekoa We could consider changing the interface to `execute_agent()`. Then we wouldn't need to pass a command builder into the `BruteForceExploiter`, the client would just take care of it. If we're not using `execute_detached()` for anything other than executing the agent, perhaps this is a worthwhile change (but in the next PR). Thoughts?

Reply:
> From the BruteForceExploiter's perspective, it's a necessity that the command be detached. How that gets achieved is up to the client.

What if we brute-force a process that's unable to detach his children? I remember having to sweat a lot to figure out a workaroud for this on one of the exploiters, I'm not sure we want to enforce this requirement in the interface

Reply:
We were operating under the assumption that the command could be anything, but from the perspective of the `BruteForceExploiter`, it only expects the command to run the agent binary that gets copied. Do we expect `IRemoteAccessClient` to only be used by the `BruteForceExploiter`? If so, the change makes sense to me. If the interface changes to `execute_agent()`, we would still want some mechanism to build the command-line arguments for the agent, as that could change